### PR TITLE
Clarify use of node-role labels within Kubernetes

### PR DESCRIPTION
**Pull request by smarterclayton**
https://github.com/kubernetes/enhancements/issues/1143

Node role labels were not intended for use internal to Kubernetes components because they are not required for conformant clusters. Replace uses with implementations that match the linked KEP.

/kind feature

```release-note
Introduce `node.kubernetes.io/exclude-balancer` and `node.kubernetes.io/exclude-disruption` labels in alpha to prevent cluster deployers from being dependent on the optional `node-role` labels which not all clusters may provide.
```

### CODE REVIEW

File: pkg/controller/nodelifecycle/node_lifecycle_controller.go [hunk 1]
```diff
@@ -712,8 +713,8 @@ func (nc *Controller) monitorNodeHealth() error {
 			continue
 		}
 
-		// We do not treat a master node as a part of the cluster for network disruption checking.
-		if !system.IsMasterNode(node.Name) {
+		// Some nodes may be excluded from disruption checking
+		if !isNodeExcluded(node) {
```
Comment (liggitt) [inline pkg/controller/nodelifecycle/node_lifecycle_controller.go]: more specific of a name, like `isNodeExcludedFromDisruptionChecks`?

File: pkg/controller/nodelifecycle/node_lifecycle_controller.go [hunk 2]
```diff
@@ -804,6 +805,42 @@ func (nc *Controller) monitorNodeHealth() error {
...
+			return true
+		}
+	}
+	return false
+}
+
+// legacyIsMasterNode returns true if given node is a registered master according
+// to the logic historically used for this function. This code path is deprecated
+// and the node disruption exclusion label should be used in the future.
+// DEPRECATED: Will be removed in 1.19
```
Comment (liggitt) [inline pkg/controller/nodelifecycle/node_lifecycle_controller.go]: add a comment that this should explicitly not be updated to check the node role label

File: pkg/controller/service/service_controller.go [hunk 3]
```diff
@@ -60,13 +60,19 @@ const (
 	minRetryDelay = 5 * time.Second
 	maxRetryDelay = 300 * time.Second
 
-	// LabelNodeRoleMaster specifies that a node is a master
-	// It's copied over to kubeadm until it's merged in core: https://github.com/kubernetes/kubernetes/pull/39112
+	// LabelNodeRoleMaster specifies that a node is a master. This label is deprecated and
+	// only considered when the LegacyNodeRoleBehavior feature gate is on.
 	LabelNodeRoleMaster = "node-role.kubernetes.io/master"
```
Comment (liggitt) [inline pkg/controller/service/service_controller.go]: can we make this private or inline to avoid picking up new references?

File: pkg/controller/service/service_controller.go [hunk 4]
```diff
@@ -610,13 +616,18 @@ func getNodeConditionPredicate() corelisters.NodeConditionPredicate {
 			return false
 		}
 
-		// As of 1.6, we will taint the master, but not necessarily mark it unschedulable.
-		// Recognize nodes labeled as master, and filter them also, as we were doing previously.
-		if _, hasMasterRoleLabel := node.Labels[LabelNodeRoleMaster]; hasMasterRoleLabel {
-			return false
+		if utilfeature.DefaultFeatureGate.Enabled(kubefeatures.LegacyNodeRoleBehavior) {
```
Comment (liggitt) [inline pkg/controller/service/service_controller.go]: add a comment for what release this is eligible to be removed

File: pkg/features/kube_features.go [hunk 5]
```diff
@@ -488,7 +500,9 @@ var defaultKubernetesFeatureGates = map[featuregate.Feature]featuregate.FeatureS
 	CPUManager:                          {Default: true, PreRelease: featuregate.Beta},
 	CPUCFSQuotaPeriod:                   {Default: false, PreRelease: featuregate.Alpha},
 	TopologyManager:                     {Default: false, PreRelease: featuregate.Alpha},
+	LegacyNodeRoleBehavior:              {Default: true, PreRelease: featuregate.Alpha},
```
Comment (liggitt) [inline pkg/features/kube_features.go]: move down to the `features that enable backwards compatibility but are scheduled to be removed` section

File: test/e2e/framework/metrics_util.go [hunk 6]
```diff
@@ -0,0 +1,856 @@
+/*
```
Comment (liggitt) [inline test/e2e/framework/metrics_util.go]: bad rebase? not sure where this file came from

File: test/e2e/system/system_utils.go [hunk 7]
```diff
@@ -20,9 +20,9 @@ import (
 	"strings"
 )
 
-// IsMasterNode returns true if given node is a registered master.
-// TODO: find a better way of figuring out if given node is a registered master.
-func IsMasterNode(nodeName string) bool {
+// DeprecatedIsMasterNode returns true if given node is a registered master.
+// DEPRECATED: use a label selector provided by test initialization.
+func DeprecatedIsMasterNode(nodeName string) bool {
```
Comment (liggitt) [inline test/e2e/system/system_utils.go]: discourage even more? `DeprecatedMightBeMasterNode`? also add a comment explicitly saying this should not be updated to use the node role label

File: pkg/controller/nodelifecycle/node_lifecycle_controller.go [hunk 8]
```diff
@@ -804,6 +805,42 @@ func (nc *Controller) monitorNodeHealth() error {
 	return nil
 }
 
+// labelNodeDisruptionExclusion is a label on nodes that controls whether they are
+// excluded from being considered for disruption checks by the node controller.
+const labelNodeDisruptionExclusion = "node.kubernetes.io/exclude-disruption"
```
Comment (timothysc) [inline pkg/controller/nodelifecycle/node_lifecycle_controller.go]: Shouldn't this be in a publicly exported place?

File: pkg/controller/nodelifecycle/node_lifecycle_controller.go (hunk 2 above)
Comment (timothysc) [reply pkg/controller/nodelifecycle/node_lifecycle_controller.go]: Do we have a central doc or exported library on reserved labels?

**Review by timothysc — COMMENTED** [review]
/cc @neolit123 @fabriziopandini

**Comment by smarterclayton** [comment]
> Do we have a central doc or exported library on reserved labels?

The well known labels are part of core/v1.  Things that are new should be added there, although components are not required to consume those constants.  I think we should add labels once they hit beta to well_known.

File: pkg/controller/nodelifecycle/node_lifecycle_controller.go (hunk 8 above)
Comment (smarterclayton) [reply pkg/controller/nodelifecycle/node_lifecycle_controller.go]: Until it hits beta, I would say no.  Referencing constants across multiple concrete domains (controller manager / rest of kube) would be a bug.

File: test/e2e/system/system_utils.go (hunk 7 above)
Comment (smarterclayton) [reply test/e2e/system/system_utils.go]: OK

File: test/e2e/framework/metrics_util.go (hunk 6 above)
Comment (smarterclayton) [reply test/e2e/framework/metrics_util.go]: Bad rebase, they refactored under me

File: pkg/controller/service/service_controller.go (hunk 3 above)
Comment (smarterclayton) [reply pkg/controller/service/service_controller.go]: already on the way out

**Review by neolit123 — COMMENTED** [review]
deprecating IsMasterNode seems like the correct way to go.

File: pkg/controller/service/service_controller.go [hunk 9]
```diff
@@ -60,13 +60,19 @@ const (
...
 	maxRetryDelay = 300 * time.Second
 
-	// LabelNodeRoleMaster specifies that a node is a master
-	// It's copied over to kubeadm until it's merged in core: https://github.com/kubernetes/kubernetes/pull/39112
-	LabelNodeRoleMaster = "node-role.kubernetes.io/master"
-
-	// LabelNodeRoleExcludeBalancer specifies that the node should be
-	// exclude from load balancers created by a cloud provider.
-	LabelNodeRoleExcludeBalancer = "alpha.service-controller.kubernetes.io/exclude-balancer"
+	// labelNodeRoleMaster specifies that a node is a master. This label is deprecated and
```
Comment (justinsb) [inline pkg/controller/service/service_controller.go]: I don't think "deprecated" captures what you wrote in the KEP; propose "the use of this label to determine whether a node is attached to a loadbalancer is deprecated"

File: pkg/features/kube_features.go [hunk 10]
```diff
@@ -171,12 +171,24 @@ const (
 	// Enable pods to set sysctls on a pod
 	Sysctls featuregate.Feature = "Sysctls"
 
+	// owner @smarterclayton
+	// alpha: v1.16
+	//
+	// Enable legacy behavior to vary cluster functionality on the node-role.kubernetes.io labels. On by default (legacy), will be turned off in 1.18.
```
Comment (justinsb) [inline pkg/features/kube_features.go]: "cluster functionality" is ambiguous - propose we list the functionality it turns off: "Allow Service LoadBalancer to respect the node-role.kubernetes.io label when determining node inclusion."   Or maybe rephrased in human :-)

File: test/e2e/system/system_utils.go [hunk 11]
```diff
@@ -20,9 +20,9 @@ import (
 	"strings"
 )
 
-// IsMasterNode returns true if given node is a registered master.
-// TODO: find a better way of figuring out if given node is a registered master.
-func IsMasterNode(nodeName string) bool {
+// DeprecatedIsMasterNode returns true if given node is a registered master.
+// DEPRECATED: use a label selector provided by test initialization.
```
Comment (justinsb) [inline test/e2e/system/system_utils.go]: As this is now in the tests (yay!), and in the absence of a proposal on how to deal with e2e tests that act on the control plane, it feels wrong to deprecate this.

Propose we move the label to a flag, and put a warning like "prefer other mechanisms to atomically considering a node to be master or not".

File: test/test_owners.csv [hunk 12]
```diff
@@ -793,7 +793,6 @@ k8s.io/kubernetes/pkg/util/oom,vishh,0,
 k8s.io/kubernetes/pkg/util/parsers,derekwaynecarr,1,
 k8s.io/kubernetes/pkg/util/procfs,roberthbailey,1,
 k8s.io/kubernetes/pkg/util/slice,quinton-hoole,0,
-k8s.io/kubernetes/pkg/util/system,mwielgus,0,
```
Comment (justinsb) [inline test/test_owners.csv]: Deliberate?

File: pkg/controller/service/service_controller.go (hunk 4 above)
Comment (justinsb) [reply pkg/controller/service/service_controller.go]: I'd say that is premature until the replacement is rolled out & proven.

File: test/test_owners.csv (hunk 12 above)
Comment (smarterclayton) [reply test/test_owners.csv]: This package was removed.

File: test/e2e/system/system_utils.go (hunk 11 above)
Comment (smarterclayton) [reply test/e2e/system/system_utils.go]: Why is it wrong to deprecate?  Something that is broken is broken - this check doesn't even work on a subset of conformant clusters, so sig-testing definitely needs to help identify next steps.  This is being explicitly marked this way to clearly indicate that no one should add new code using this, which is something Jordan and I have seen repeatedly over the last month.

File: pkg/features/kube_features.go (hunk 10 above)
Comment (smarterclayton) [reply pkg/features/kube_features.go]: We generally don't get very descriptive in these gates.  The behavior we are removing is "using node-role label to change things".  Expectation is that people look at the use of the gates or docs, but not here.

File: pkg/controller/service/service_controller.go (hunk 9 above)
Comment (smarterclayton) [reply pkg/controller/service/service_controller.go]: Agreed, clarified.

File: test/e2e/system/system_utils.go (hunk 11 above)
Comment (smarterclayton) [reply test/e2e/system/system_utils.go]: The proposal on how to act on e2e tests on the control plane is to accept labels that identify masters as a parameter (defaulting to existing), if no masters are present the test should skip or continue (depending), and the use of that flag is considered what this is deprecated in favor of.

**Comment by smarterclayton** [comment]
Rebased, clarified the comment

**Comment by timothysc** [comment]
/assign @justinsb 
^ as he had the most to say from SCL.

File: pkg/controller/service/service_controller.go [hunk 13]
```diff
@@ -60,13 +60,20 @@ const (
...
-	LabelNodeRoleExcludeBalancer = "alpha.service-controller.kubernetes.io/exclude-balancer"
+	// labelNodeRoleMaster specifies that a node is a master. The use of this label within the
+	// controller is deprecated and only considered when the LegacyNodeRoleBehavior feature gate
+	// is on.
+	labelNodeRoleMaster = "node-role.kubernetes.io/master"
+
+	// labelNodeRoleExcludeBalancer specifies that the node should be
+	// exclude from load balancers created by a cloud provider. This label is honored starting
+	// in 1.16 when the ServiceNodeExclusion gate is on.
+	labelNodeRoleExcludeBalancer = "node.kubernetes.io/exclude-balancer"
```
[the remark this answers is not in the document]
Comment (thockin) [reply pkg/controller/service/service_controller.go]: Put another way:

Assume:
* Node N, with this "exclude LB" label
* Service S1, type=LoadBalancer, externalTrafficPolicy=Cluster
* Pod P1 which matches S1, running on N
* Service S2, type=LoadBalancer, externalTrafficPolicy=Local
* Pod P2 which matches S2, running on N

Consider this situation in cloud A, which uses 2-hop load-balancing and on cloud B, which uses direct-to-pod load-balancing.

On cloud A (2-hop):
* P1 is reachable by the S1 LB via a different node hop.
* Is P2 *required*, *allowed*, or *required not* to be reachable by the S2 LB (via a hop on node)?

On cloud B (direct):
* P1 is reachable by the S1 LB directly.
* Is P2 *required*, *allowed*, or *required not* to be reachable by the S2 LB (directly)?

We have a weird intersection of 2 features, and I'd like to get agreement on the expected semantics before documenting this.  I will open a new sig-network issue.

[the remark this answers is not in the document]
Comment (thockin) [reply pkg/controller/service/service_controller.go]: https://github.com/kubernetes/kubernetes/issues/81178

[the remark this answers is not in the document]
Comment (smarterclayton) [reply pkg/controller/service/service_controller.go]: You always find the most exciting complexities.  I blame networking.

`node.kubernetes.io/exclude-from-external-load-balancers` sounds fine to me if thats' the desired clarification.  I think we should define what we mean, if we find the need to document multiple ways we can.  After all, we got in this spot by being too general.  I think we should be as specific as the use case and no less specific.

**Comment by smarterclayton** [comment]
Updated with the new name.

**Review by thockin — COMMENTED** [review]
/lgtm
